### ISSUE TITLE
It feels like right is treated as left 

### ISSUE THREAD

**Issue by iaomw** [robb/Cartography#148]
Hi, I am using the 0.5.0 release, and I am doing something like that below. I really like the syntax, but It feels like right is treated as left, but left works fine. Any help 

layout(label, self.nameLabel) { thisView, anotherView in
            
            thisView.right == thisView.superview!.right - 33
            thisView.centerY == anotherView.centerY
 }

**Comment by robb** [comment]
You shouldn't have to, can you share some code maybe?

**Comment by iaomw** [comment]
lazy var nameLabel:UILabel = {
        
    let label = UILabel()
    
    label.font = someFont
    label.textColor = someColor
    
    self.contentView.addSubview(label)
    
    layout(label) { thisView in
        
        thisView.left == thisView.superview!.left + 33
        thisView.bottom == thisView.superview!.centerY - 1
    }
    
    layout(self.contentView, label) { thisView, anotherView in
        
        thisView.height == anotherView.height + 46
        thisView.width == thisView.superview!.width  // There will be problem without this line
    }

    return label
    
    } ()

**Comment by iaomw** [comment]
But I have another cell works well with pure NSLayoutConstraint without bug and warning
```
lazy var nameLabel:UILabel = {
        
    let label = UILabel()
    
    label.setTranslatesAutoresizingMaskIntoConstraints(false)
    
    label.font = MEFont.fontWithName(.RestaurantTitleInList)

    self.contentView.addSubview(label)
    
    let leftConstraint = NSLayoutConstraint(item: label, attribute: .Left, relatedBy: .Equal, toItem: self.contentView, attribute: .Left, multiplier: 1, constant: padding)
    let centerYConstraint = NSLayoutConstraint(item: label, attribute: .CenterY, relatedBy: .Equal, toItem: self.contentView, attribute: .CenterY, multiplier: 1, constant: 0)
    
    let heightConstraint = NSLayoutConstraint(item: self.contentView, attribute: .Height, relatedBy: .Equal, toItem: label, attribute: .Height, multiplier: 1, constant: 38)
    
    self.contentView.addConstraints([leftConstraint, centerYConstraint, heightConstraint])
    
    return label
    
    } ()
```

**Comment by robb** [comment]
Could you try this:

```Swift
layout(label) { label in
    label.left   == label.superview!.left + 33
    label.bottom == label.superview!.centerY - 1
}

layout(self.contentView, label) { contentView, label in
    label.height == contentView.height + 46
}
```

I think what's happening is that the `contentView` relies on its resizing masks getting translated into constraints automatically, which [Cartography will disable for the left hand side of every constraint](https://github.com/robb/Cartography#usage).

**Comment by iaomw** [comment]
Thanks, I have tried your code. There is no crash nor warning, but the cell lost it's height, then it's using the default height provided by TableVIew.

**Comment by robb** [comment]
Hey, sorry for not checking back in sooner.

Does the problem still persist?

**Comment by iaomw** [comment]
Yes, I still need to set the width if using the self-sizing cell. And sometimes I also need to set the left of contentView to make sure it's not positioned to some strange place in the cell.
```
layout(self.contentView, label, replace: self.sizeC4ContentView) { thisView, anotherView in
   
            thisView.height == anotherView.height + 45
            thisView.width == thisView.superview!.width
            thisView.left == thisView.superview!.left
}
```

**Comment by robb** [comment]
Does it help if you don't have the `contentView` as the left hand side argument of the `==` operator? Btw,  I recommend keeping the names consistent with the parent scope, so you don't need to translate between `thisView` and `contentView` all the time :v:

**Comment by iaomw** [comment]
Sorry, I miss the point. If I don't put the `contentView` on the left of `==` operator, how could I update the height of contentView based on the height of label?

**Comment by robb** [comment]
Does `contentView.height == label.height - 45` not work?

**Comment by iaomw** [comment]
Then the contentView become very small, and everything get weird. Why `-` the height of `contentView` should be larger? I don't understand.  😵😵😵

**Comment by vfn** [comment]
@iaomw, set all the constrains of you label and at the end set the contentView's bottom to match the label's bottom

```Swift
self.contentView.addSubview(label)
constrain(label) { view in
            let superview = view.superview!
            
            view.top == superview.top
            view.left == superview.left
            view.right == superview.right
            
            superview.bottom == view.bottom + 45
}
```

**Comment by iaomw** [comment]
Thanks @vfn The height of `contentView` really work in this way. But the the width of `contentView` is still not correct, it should go with the width of cell. 
So everything in `contentView` should adjust in this style. I will rewrite other AL code to see how it works.

**Comment by vfn** [comment]
Just make sure that the content view has the correct constraints are in place

```Swift
        self.contentView.translatesAutoresizingMaskIntoConstraints = false
        constrain(self.contentView) { view in
            let superview = view.superview!
            
            view.top == superview.top
            view.left == superview.left
            view.right == superview.right
        }
```

**Comment by iaomw** [comment]
@vfn This method does resolved the size problem of contentView, but the label will not resize itself properly. It's a muti-line label, but it will appear to be a single line label after that.

**Comment by robb** [comment]
Does this help? From http://johnszumski.com/blog/auto-layout-for-table-view-cells-with-dynamic-heights

> Another less obvious pain point is using multi-line UILabels in a layout. I expected that if a label's text had more than one line of content at its current width, that the label's intrinsicContentSize would simply expand to a height that fits the entire text string. Assuming appropriate constraints existed to allow the label to grow to that height, it would be fully visible in the layout. Unfortunately, Auto Layout instead will only render the label with one line of text and truncate the rest. The fix is to set preferredMaxLayoutWidth to the width of your label, which allows the layout engine to correct compute the full height. Having to know the proper width ahead of time is a major restriction, and I created a UILabel subclass that keeps its preferredMaxLayoutWidth updated as its bounds change:

```objc
- (void)layoutSubviews {
    [super layoutSubviews];
     
    self.preferredMaxLayoutWidth = CGRectGetWidth(self.bounds);
                 
    [super layoutSubviews];
}
```

**Comment by iaomw** [comment]
Thanks, I know that. But I have different elements for touching and normal. And they also have different sizes and positions for touching and normal. So, basically the width is also not a const, I cannot use `preferredMaxLayoutWidth`, it should be ruled by other AL code like constraint.

**Comment by vfn** [comment]
@iaomw, you'll need to inspect your view hierarchy and try to understand what is going wrong there. There is probably one or more constraints missing. Tools like http://revealapp.com/ can help you with that

Alternatively, you can create a sample project and we can try to help you by looking at your code. I do have a couple of self sizing cells on the project I'm using Cartography on and They work as expected.

**Comment by WestFlow127** [comment]
I have a similar issue with height either overlapping the cells below it or the cell's height being too big when not overlapping.

    constrain(self.contentView, messageLabel, bottomView, userImageView) { contentView, msgLabel, bottomView, userImageView in
            
       bottomView.top == msgLabel.bottom
       bottomView.right == msgLabel.rightMargin
       bottomView.left == userImageView.right + Geometry.messageLabelLeftOffset
            
       contentView.bottom == bottomView.bottom // very tall cell is overlapping cells below with this line
       //  contentView.bottom.height == msgLabel.height + 20 // cell is twice as tall,  leaving large empty gap, but not overlapping
    }

heightForRowAt is UITableView.automaticDimesnion and the bottomView is the bottom-most view of TableViewCell.

Anyone have some suggestions as to what I should do differently?
